solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java

```diff
@@ -139,30 +138,10 @@ public static String getCollectionPathRoot(String coll) {
    * only a replica is updated
    */
   public DocCollection copyWith(PerReplicaStates newPerReplicaStates) {
```

Comment:
I assume we will need to modify `getSlices()` so it would return the correct "view" of slices from replica states too? Anyway, this is probably still WIP ! 😊

Reply:
Wy would slices change when PRS change?

Reply:
Based on the old code, the modified replicas would be read from the newPerReplicaStates provided to construct a list as `modifiedShards` (which each Slice  value instead contains a map of replicas, which info relies on the input `newPerReplicaStates`).

Since we are not constructing a new DocCollection here and instead returning the same `this` instance, how can we ensure the such DocCollection instance getSlice calls return the updated replica info from `newPerReplicaStates`?

Reply:
when per-replica states change the slices remain the exactly the same

Reply:
By following the old code logic, it appears to be necessary to update the "replicas" stored within "Slice"?

For example if someone call docCollection.getSlice("shard1").getReplica("core_node_2"), the `Replica` stored in the `Slice` might not reflect the latest one from `newPerReplicaStates`  since they are all direct field and map lookup afaik :)

Reply:
The `DocCollection` , all the Slices & Replicas share the same `PrsSupplier`. When the PRS data changes, it's immediately visible to all the objects at once

Reply:
The test here https://github.com/cowpaths/fullstory-solr/commit/8a83fdfc20d4ae7dc832217ed8533aca3c0bb4ee fails with the current branch as Slice and Replica are direct access to read from `Map` (as in https://github.com/cowpaths/fullstory-solr/blob/ef15b8615d01b9a4c00731c489c57dfe36621201/solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java#L233 and  https://github.com/cowpaths/fullstory-solr/blob/ef15b8615d01b9a4c00731c489c57dfe36621201/solr/solrj/src/java/org/apache/solr/common/cloud/Slice.java#L304). Unless the `Map` is a specialized version that knows to read from `PrsSupplier`?

Reply:
The test was wrong. I have fixed it there.

Reply:
Thanks i see what you mean now! 😅  I didn't notice the how the information was shared/mutated by the same `prsSupplier`.

I guess what confused me was that I have always assumed Slice/Replica to be immutable so it was unexpected on how the flow now works.

I do understand we are trying to avoid making copies here by reusing the existing Slice/Shard instances, and it might then become inevitable to mutate them. 

Wondering whether the existing design (immutable/making copies of the replica/shard on update) has caused any performance concern? Being immutable does have certain advantages - such as thread safety and ease to read the flow - considered that i just missed the shared prsSupplier details, this might not be obvious of how changes are shared, as such linkage is built at the ctor of `DocCollection`, which is outside of the scope of Slice/Replica

Reply:
The objective is to closely map the new design with the way objects are created/modified/used

Before PRS, the `state.json` contained everything. So, once the `state.json` is loaded `DocCollection`, `Slice`, `Replica` Objects  remain immutable. After PRS, that assumption is no longer true. The optimisation that we did in ZK should reflect in the in memory structure as well.

> ... which is outside of the scope of Slice/Replica

Yes, the PRS data is itself outside the scope of `Slice`/`Replica`